lib/THC/THCTensorMode.cuh

```diff
@@ -0,0 +1,282 @@
+
+  // Our last shared memory buffer is used to track indices
+  __shared__ unsigned int imem[Power2Size];
+
+  // initialize the indices buffer such that imem[i] = i
+  imem[tidx] = tidx;
+  imem[stidx] = stidx;
+  __syncthreads(); // barrier for both the scan and the imem initialization
+
+  // At this point, we need to find the maximum element in the cmem buffer.
+  // This element will represent the count (-1) of the mode. Because of the
+  // way we have set up the problem, the index where this mode occurs will
+  // also be the location of the mode value in the sorted array, e.g.
+  //
+  // smem = [0, 0, 1, 1, 1, 2]
+  // cmem = [0, 1, 0, 1, 2, 0]
+  //                     ^
+  //                     maximum value, also aligned with mode = 1
+  //
+  // We perform a block wide max-reduction of the cmem buffer, and bring imem
+  // along with it.
+  //
+  // Loop 1 (Power2Size = offset = 4):
+  //
+  // (0, 4) --> cmem[4] = 2 > cmem[0] = 0, so update cmem[0], imem[0]
+  // (1, 5) --> cmem[5] = 0 <= cmem[1] = 1, do nothing
+  //
+  // Now:         0  1  2  3  4  5
+  //      cmem = [2, 1, 0, 1, 2, 0]
+  //      imem = [4, 1, 2, 3, 4, 5]
+  //
+  // Loop 2 (offset = 2)
+  //
+  // (0, 2) --> cmem[2] == 0 <= cmem[0] = 2, do nothing
+  // (1, 3) --> cmem[3] == 1 <= cmem[1] = 1, do nothing
+  //
+  // Now:         0  1  2  3  4  5
+  //      cmem = [2, 1, 2, 1, 2, 0]
+  //      imem = [4, 1, 4, 3, 4, 5]
+  //
+  // Loop 3 (offset = 1)
+  //
+  // (0, 1) --> cmem[1] == 1 <= cmem[0] = 2, do nothing
+  //
+  // So at the end at cmem[0] we have the maximum count = 2, and the
+  // corresponding index = 4
+#pragma unroll
+  for (unsigned int offset = Power2Size / 2; offset > 0; offset >>= 1) {
+    if (tidx < offset && tidx + offset < sliceSize) {
+      // Note that we could do >= as well. We use >, so that we pick the
+      // earliest maximum value in the sequence in case of ties. This will
+      // result in picking the smallest value for the mode, i.e. if both
+      // 3 and 4 occur the same number of times in the input, and their count
+      // is the mode, then we return 3. This mimics the behavior of CPU-Torch
+      if (cmem[tidx + offset] > cmem[tidx]) {
+        cmem[tidx] = cmem[tidx + offset];
+        imem[tidx] = imem[tidx + offset];
+      }
+    }
+    __syncthreads();
+  }
+
+  // Store the mode in shared memory for use in finding the mode in the input slice
+  __shared__ T  mode;
+
+  // Given the above constraints, the mode is the value at the maximum index in the segmented scan
+  if (tidx == 0) {
+    mode = smem[imem[0]];
+  }
+  __syncthreads(); // broadcast mode
+
+  // Finally, we need to find the "an" index of the mode in the input Tensor. The API does
+  // not constrain which index we pick, so it can be any of the indices that contain the mode.
+  // We will do a reduction to find the index. First, we mark indices that are equal to the mode,
+  // i.e bmem[i] = true if input[i] == mode
+  if (tidx < sliceSize) {
```

Comment:
Consider using `reduceBlock` twice, or adding functionality to `reduceBlock` where you can provide multiple values instead of duplicating more reduction code; e.g.,

```
int indices[2] = { ... };
reduceBlock<int, Min<int>, 2>(imem, sliceSize, indices, Min<int>, MAX_INT);
```

all you need is a min reduction on an integer index value, where the value is the current `tidx` or `stidx` for threads that have a mode value, and MAX_INT otherwise.

Then, if we add the warp shuffle optimization to `reduceBlock`, this code will get faster as well.